transaction/core/src/tx.rs

```diff
@@ -237,6 +243,44 @@ pub struct TxIn {
     /// Prost only works with Vec.
     #[prost(message, repeated, tag = "2")]
     pub proofs: Vec<TxOutMembershipProof>,
+
+    /// Any rules associated to this input, per MCIP #31
+    #[prost(message, tag = "3")]
+    pub input_rules: Option<InputRules>,
+}
+
+impl TxIn {
+    /// This is the digest of the TxIn which is signed per MCIP #31 if there
+    /// are input rules present.
+    ///
+    /// See MCIP #31 for rationale -- by not signing the whole TxPrefix, we
+    /// allow that someone can create this signature who does not have the
+    /// whole TxPrefix.
+    ///
+    /// The membership proofs are not signed, because it is useful to allow that
+    /// someone later may update those proofs. See MCIP #31 for discussion.
+    pub fn signed_digest(&self) -> Option<[u8; 32]> {
```

Comment:
calling this out for reviewers:

in the latest revision, this is the digest that is signed, and the `input_rules.digest()` thing went away.

the idea is, sign the entire `TxIn`, but remove the merkle proofs first. The reason for this is that there are tx-wide rules about merkle proofs having matching trust roots, so it may be necessary for the recipient of a signed input to update these proofs without breaking the signature.

Reply:
I believe that this is the largest thing here that could possibly be signed, everything else in the signed contingent input is either implied, or is supposed to be secret from the consensus enclave

Reply:
Yeah that's a good point that the trust roots will change when the matching contingent inputs are combined.